File: .github/scripts/update_posts.py
import requests
from bs4 import BeautifulSoup
from github import Github, Auth
import os
import re

TOKEN = os.environ.get('GH_TOKEN')
URL = "https://hunj.dev"
HEADER = "### Recent Blog Posts"
REPO = 'hunj/hunj'
END = "Read more"
REGEX = rf"{HEADER}[\s\S]*?(?={END})"

req = requests.get(URL)
soup = BeautifulSoup(req.content, "html.parser")
posts = [HEADER]

for post in soup.select('article.post'):
    title = post.select_one('h2.post-title').text.strip()
    path = post.select_one('a.post-title-link')['href']
    text = f"- [{title}]({URL}{path})"
    posts.append(text)

posts.append('\n')
posts_text = '\n'.join(posts)

shithub = Github(auth=Auth.Token(TOKEN))
repo = shithub.get_repo(REPO)
readme = repo.get_readme()

new_content = re.sub(REGEX, posts_text, readme.decoded_content.decode())
repo.update_file(readme.path, 'Update recent blog posts', new_content, readme.sha)

File: .github/scripts/update_posts.rb
require 'httparty'
require 'nokogiri'
require 'octokit'

# Scrape blog posts from the website
url = "https://hunj.dev"
response = HTTParty.get(url)
parsed_page = Nokogiri::HTML(response.body)
posts = parsed_page.css('.flex.flex-col.rounded-lg.shadow-lg.overflow-hidden')

# Generate the updated blog posts list (top 5)
posts_list = ["\n### Recent Blog Posts\n\n"]
posts.first(5).each do |post|
  title = post.css('h2.post-title').text.strip
  link = "https://#{url}#{post.at_css('a')[:href]}"
  posts_list << "* [#{title}](#{link})"
end

# Update the README.md file
client = Octokit::Client.new(access_token: ENV['GITHUB_TOKEN'])
repo = ENV['GITHUB_REPOSITORY']
readme = client.readme(repo)
readme_content = Base64.decode64(readme[:content]).force_encoding('UTF-8')

# Replace the existing blog posts section
posts_regex = /### Recent Blog Posts\n\n[\s\S]*?(?=<\/td>)/m
updated_content = readme_content.sub(posts_regex, "#{posts_list.join("\n")}\n")

client.update_contents(repo, 'README.md', 'Update recent blog posts', readme[:sha], updated_content)
